File: python/markdown/markdown.py
import re

BOLD = re.compile(r"__(.*?)__")
ITALIC = re.compile(r"_(.*?)_")
TITLE = re.compile(r"^(#{1,6}) (.*)")
OPENING_LIST = "<ul>"
CLOSING_LIST = "</ul>"


def parse(markdown):
    markdown = re.sub(BOLD, r"<strong>\1</strong>", markdown)
    markdown = re.sub(ITALIC, r"<em>\1</em>", markdown)

    result = []
    for line in markdown.split("\n"):
        is_title, title = parse_heading(line)
        if is_title:
            result.append(title)
        elif line.startswith("* "):
            item = "<li>{item}</li>".format(item=line[2:])
            if result and result[-1] == CLOSING_LIST:
                result[-1] = item
            else:
                result.append(OPENING_LIST)
                result.append(item)
            result.append(CLOSING_LIST)
        else:
            result.append("<p>{line}</p>".format(line=line))

    return "".join(result)


def parse_heading(line):
    title = re.match(TITLE, line)
    if title:
        return True, "<h{level}>{content}</h{level}>".format(
            level=len(title.group(1)), content=title.group(2)
        )

    return False, ""

File: elixir/markdown/lib/markdown.ex
defmodule Markdown do
  @tags [%{reg: ~r/__([^_]+)__/, tag: "strong"}, %{reg: ~r/_([^_]+)_/, tag: "em"}]

  @doc """
    Parses a given string with Markdown syntax and returns the associated HTML for that string.

    ## Examples

    iex> Markdown.parse("This is a paragraph")
    "<p>This is a paragraph</p>"

    iex> Markdown.parse("#Header!\n* __Bold Item__\n* _Italic Item_")
    "<h1>Header!</h1><ul><li><em>Bold Item</em></li><li><i>Italic Item</i></li></ul>"
  """
  @spec parse(String.t()) :: String.t()
  def parse(m) do
    String.split(m, "\n")
    |> Enum.map_join(&process/1)
    |> process_tags()
    |> wrap_lists()
  end

  defp process("#" <> t), do: process_header(t, 1)
  defp process("* " <> t), do: "<li>#{t}</li>"
  defp process(t), do: "<p>#{t}</p>"

  defp process_header(" " <> t, level), do: "<h#{level}>#{t}</h#{level}>"
  defp process_header("#" <> t, level), do: process_header(t, level + 1)

  defp process_tags(raw) do
    Enum.reduce(@tags, raw, fn rep, acc ->
      String.replace(acc, rep.reg, "<#{rep.tag}>\\1</#{rep.tag}>")
    end)
  end

  defp wrap_lists(l) do
    String.replace(l, "<li>", "<ul><li>", global: false)
    |> String.replace_suffix("</li>", "</li></ul>")
  end
end
